### PR TITLE
112 handle cascade deletes

### PR DESCRIPTION
**Pull request by njuguna-n**
# Description

Add cascade deletes to models and use post hook to delete base models

[medic/sync#112
](https://github.com/medic/cht-sync/issues/112)

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: tests/e2e-test.spec.js [hunk 1]
```diff
@@ -93,27 +94,26 @@ describe('Main workflow Test Suite', () => {
...
     const person = contacts().find(contact => contact.type === 'person' && !contact._deleted);
 
-    const preDelete = await client.query(`SELECT * FROM ${pgSchema}.persons where _id = $1`, [person._id]);
+    const preDelete = await client.query(`SELECT * FROM ${pgSchema}.persons where uuid = $1`, [person._id]);
     expect(preDelete.rows.length).to.equal(1);
 
     await deleteDoc(person);
     await delay(6); // wait for CHT-Sync
     await delay(6); // wait for DBT
     await delay(6); // wait for DBT
```
Comment (lorerod) [inline tests/e2e-test.spec.js:110]:
```suggestion
    await delay(6); // wait for CHT-Sync
    await delay(12); // wait for DBT
```

File: tests/e2e-test.spec.js [hunk 2]
```diff
@@ -93,27 +94,26 @@ describe('Main workflow Test Suite', () => {
...
 
-    const modelContactResult = await client.query(`SELECT * FROM ${pgSchema}.contacts where _id = $1`, [person._id]);
-    expect(modelContactResult.rows[0]._deleted).to.equal(true);
```
Comment (lorerod) [inline tests/e2e-test.spec.js:113]: Does the contact row get deleted at some point?

File: tests/e2e-test.spec.js [hunk 3]
```diff
@@ -93,27 +94,26 @@ describe('Main workflow Test Suite', () => {
     const pgTableContact = await client.query(`SELECT * from ${PGTABLE} where _id = $1`, [contact._id]);
     expect(pgTableContact.rows[0]._deleted).to.equal(true);
     await delay(6); // wait for DBT
-    const modelContactResult = await client.query(`SELECT * FROM ${pgSchema}.contacts where _id = $1`, [contact._id]);
-    expect(modelContactResult.rows.length).to.equal(1);
-    expect(modelContactResult.rows[0]._deleted).to.equal(true);
+    const modelContactResult = await client.query(`SELECT * FROM ${pgSchema}.contacts where uuid = $1`, [contact._id]);
+    expect(modelContactResult.rows.length).to.equal(0);
```
Comment (lorerod) [inline tests/e2e-test.spec.js:98]: What would happen if the contact deleted was of the type ´person´?

**Review by lorerod — CHANGES_REQUESTED** [review]

File: tests/dbt/package/models/contacts/contacts.yml [hunk 4]
```diff
@@ -2,13 +2,56 @@ version: 2
...
+    config:
+      contract:
+        enforced: true
     columns:
-      - name: _id
+      - name: uuid
+        data_type: string
+        constraints:
+          - type: foreign_key
+            expression: "{{ env_var('POSTGRES_SCHEMA') }}.contacts (uuid) ON DELETE CASCADE"
```
Comment (lorerod) [inline tests/dbt/package/models/contacts/contacts.yml:49]: Do we need something similar in `reports.yml`? The deleted reports are not propagating from `v1.medic` to `dbt.Reports`  The deletions of contacts and persons are working fine. See video:

https://github.com/medic/cht-sync/assets/21312057/4409de75-3c43-4225-b9d4-e3e7a89ff4d7

**Review by lorerod — CHANGES_REQUESTED** [review]

File: tests/e2e-test.spec.js (hunk 2 above)
Comment (njuguna-n) [reply tests/e2e-test.spec.js:113]: Yes, we use a `post_hook` to [delete](https://github.com/medic/cht-sync/blob/e4408e1f3a8da2be3608c1f10bcdd3733d6a8785/tests/dbt/package/models/contacts/contacts.sql#L6-L7) any rows with the `_deleted` flag set to true

File: tests/e2e-test.spec.js (hunk 3 above)
Comment (njuguna-n) [reply tests/e2e-test.spec.js:97]: We would expect the contact to be deleted from both the `contact` and `person` tables.

File: tests/e2e-test.spec.js (hunk 2 above)
Comment (lorerod) [reply tests/e2e-test.spec.js:113]: How can I see this when testing?

File: tests/dbt/package/models/contacts/contacts.yml (hunk 4 above)
Comment (njuguna-n) [reply tests/dbt/package/models/contacts/contacts.yml:49]: After further testing we have confirmed with @lorerod that deletions for reports also work but she had a weird bug where a report was deleted in the CHT but that was not propagated to postgres. We will try and have reproducible steps for that and create an issue once we have those.

**Review by lorerod — APPROVED** [review]
Thank you @njuguna-n deletes are working! 
As you mentioned in [comment](https://github.com/medic/cht-sync/pull/118/files#r1665910787) I will keep working on reproducing the weird bug.

File: tests/dbt/package/models/contacts/contacts.yml (hunk 4 above)
Comment (lorerod) [reply tests/dbt/package/models/contacts/contacts.yml:49]: Okay, I finally figured it out. The report still exists in the database. If I log in as an admin, I can see it.
What happened was that the report's person was deleted before the report. I wonder if this is expected in the CHT. I will ask in another thread.
Here is the video reproducing:

https://github.com/medic/cht-sync/assets/21312057/9a462d61-6189-45da-aeb7-778ee5ca5f99

**Pull request merged**